spec/grape/exceptions/body_parse_errors_spec.rb

```diff
@@ -0,0 +1,86 @@
+require 'spec_helper'
+
+describe Grape::Exceptions::ValidationErrors do
+  context 'api with rescue_from :all handler' do
+    subject { Class.new(Grape::API) }
+    before {
+      subject.format :json
+      subject.rescue_from :all do |e|
+        rack_response 'message was processed', 400
+      end
+      subject.params do
+        requires :beer
+      end
+      subject.post '/beer' do
+        'beer received'
+      end
+    }
+
+    # non-JSON body
+    body = 'test'
+
+    def app
+      subject
+    end
+
+    context 'with content_type json' do
+      headers = { 'CONTENT_TYPE' => 'application/json' }
+
+      it 'can recover from failed body parsing' do
+        post '/beer', body, headers
+        expect(last_response.status).to eq 400
+        expect(last_response.body).to eq('message was processed')
+      end
+    end
+
+    context 'with no specific content_type' do
+      headers = {}
+
+      it 'can recover from failed body parsing' do
+        post '/beer', body, headers
+        expect(last_response.status).to eq 400
+        expect(last_response.body).to eq('message was processed')
+      end
+    end
+  end
+
+  context 'api without a rescue handler' do
+    subject { Class.new(Grape::API) }
+    before {
+      subject.format :json
+      subject.params do
+        requires :beer
+      end
+      subject.post '/beer' do
+        'beer received'
+      end
+    }
+
+    # non-JSON body
+    body = 'test'
```

Comment:
This declares something in a scope that's not very clear, since who knows what `context` is :) I would just hard-code `test` below in the two places you need it and not worry about it. Same for headers.

Otherwise use `let(:body) { ...}`, which is the way RSpec tells us to declare things like this.